hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ImplicitDiscriminatorConverter.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.metamodel.mapping.internal;
+
+import org.hibernate.HibernateException;
+import org.hibernate.MappingException;
+import org.hibernate.internal.util.collections.CollectionHelper;
+import org.hibernate.metamodel.mapping.DiscriminatorConverter;
+import org.hibernate.metamodel.mapping.DiscriminatorValueDetails;
+import org.hibernate.metamodel.model.domain.NavigableRole;
+import org.hibernate.metamodel.spi.MappingMetamodelImplementor;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.type.descriptor.java.JavaType;
+
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.util.Locale.ROOT;
+
+/**
+ * @author Steve Ebersole
+ */
+public class ImplicitDiscriminatorConverter<O,R> extends DiscriminatorConverter<O,R> {
+	private final NavigableRole discriminatorRole;
+	private final MappingMetamodelImplementor mappingMetamodel;
+	private final Map<Object, DiscriminatorValueDetails> detailsByValue;
+	private final Map<String,DiscriminatorValueDetails> detailsByEntityName;
+
+	public ImplicitDiscriminatorConverter(
+			NavigableRole discriminatorRole,
+			JavaType<O> domainJavaType,
+			JavaType<R> relationalJavaType,
+			Map<Object,String> explicitValueMappings,
+			MappingMetamodelImplementor mappingMetamodel) {
+		super( discriminatorRole.getFullPath(), domainJavaType, relationalJavaType );
+		this.discriminatorRole = discriminatorRole;
+		this.mappingMetamodel = mappingMetamodel;
+
+		if ( CollectionHelper.isNotEmpty( explicitValueMappings ) ) {
+			throw new MappingException( String.format(
+					ROOT,
+					"Encountered explicit ANY discriminator mappings (%s)",
+					discriminatorRole.getFullPath()
+			) );
+		}
+
+		this.detailsByValue = CollectionHelper.concurrentMap( 8 );
+		this.detailsByEntityName = CollectionHelper.concurrentMap( 8 );
+	}
+
+	@Override
+	public DiscriminatorValueDetails getDetailsForDiscriminatorValue(Object value) {
+		if ( value instanceof String incoming ) {
+			final String entityName = mappingMetamodel.getImportedName( incoming );
```

Comment:
Maybe I'm missing something, but this implementation doesn't look quite right to me since the `detailsByEntityName` and `detailsByValue` caches aren't used.

```suggestion
			final DiscriminatorValueDetails existingDetails = detailsByValue.get( incoming );
			if ( existingDetails != null ) {
				return existingDetails;
			}
			final String entityName = mappingMetamodel.getImportedName( incoming );
```

Reply:
Oh, you mean for "read"